### PR TITLE
fix: Add /delete bidirectional relations in case of owner

### PR DESCRIPTION
**Pull request by dikshathakur3119**
Signed-off-by: dikshathakur3119 <dikshathakur@lyft.com>

### Summary of Changes
Added Bidirectional relation to add/delete in case of user and owner relationship
When we add or delete an `OWNER` relation from UI, it only deletes one way relation whereas when we do it from extractor, it add bidirectional relations, which is causing some inconsistency in neo4j graph, and creating duplicate `OWNER_OF` relations. To make this consistent, bidirectional owner relations will be created by UI API as well.
 
### Tests
No new tests added

### Documentation

_What documentation did you add or modify and why? Add any relevant links then remove this line_

### CheckList

Make sure you have checked **all** steps below to ensure a timely review.

- [ ] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [ ] PR includes a summary of changes.
- [ ] PR adds unit tests, updates existing unit tests, **OR** documents why no test additions or modifications are needed.
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
- [ ] PR passes `make test`

### CODE REVIEW

**Comment by feng-tao** [comment]
CI fails

File: metadata_service/proxy/neo4j_proxy.py
Comment (feng-tao) [inline metadata_service/proxy/neo4j_proxy.py:521]: i don't think you need to have bidirection in this query?

File: metadata_service/proxy/neo4j_proxy.py
Comment (feng-tao) [inline metadata_service/proxy/neo4j_proxy.py:1162]: could you comment on why we only delete two direction for owning resources?

Comment (feng-tao) [reply metadata_service/proxy/neo4j_proxy.py:1162]: i c, now wouldn't be an issue for bookmark/follow relationships? to do it generic, let's add bi-direction for all three relationships, and do delete for r1, r2 for all three cases? WDYT?

Comment (dikshathakur3119) [reply metadata_service/proxy/neo4j_proxy.py:1162]: Yes it would be good to have consistent relations for follow and bookmark as well. I will update it

File: metadata_service/proxy/neo4j_proxy.py
Comment (dikshathakur3119) [reply metadata_service/proxy/neo4j_proxy.py:521]: I think we would need bidirectional relation here as we are adding new relation between a user and table. Else we will again create duplicate relationships.
e.g From UI we created `Table-OWNER->User` relationship.  Now from backend we again create relationship between same table and user as 'Table-OWNER->User-OWNER_OF->Table', at this time neo4j graph will not find whole bidirectional relation hence it will create a pair of `OWNER and OWNER_OF ` relation, which will result into two `OWNER` relation and one `OWNER_OF` relationship between same user and table.
This is one case I could think of, but I feel there may be other cases as well because `MERGE` in neo4j graph looks for complete match.

**Comment by feng-tao** [comment]
lgtm, thanks @dikshathakur3119 , have you had the chance to deploy this branch to staging for a sanity test?

**Review by feng-tao — APPROVED** [review]
lgtm if done a sanity test on Lyft staging

**Comment by dikshathakur3119** [comment]
> lgtm if done a sanity test on Lyft staging

Deploying this branch on staging currently. Will merge after sanity tests.

**Comment by feng-tao** [comment]
any other issues?

**Comment by dikshathakur3119** [comment]
> any other issues?

No I did not see any other issue today. Should I merge it or wait till Monday?

File: metadata_service/proxy/neo4j_proxy.py
Comment (PaschalisDim) [inline metadata_service/proxy/neo4j_proxy.py:962]: I think those lines should include optional as in the case of deletion. If not then, the existing owners/bookmarks, will not be able to retrieve their existing ones, if they have just one relation and thus it is a breaking change.

Comment (feng-tao) [reply metadata_service/proxy/neo4j_proxy.py:962]: yeah, we should make it option and delete if so. cc @dikshathakur3119 wdyt?

Comment (dikshathakur3119) [reply metadata_service/proxy/neo4j_proxy.py:962]: Thanks @PaschalisDim for mentioning, this function is used to create query for add, delete and read, I will fix this function to return appropriate match for each of these cases.

Comment (PaschalisDim) [reply metadata_service/proxy/neo4j_proxy.py:962]: Thanks for taking this over.
@dikshathakur3119 
Creation and deletion I guess should be fine. However, when returning what exists, we need to take care to return the ones that were created prior to this MR, so that is the only thing to get altered to include all cases.
Thanks

Comment (dikshathakur3119) [reply metadata_service/proxy/neo4j_proxy.py:962]: No actually, that was reason I made this change at first place. If we do delete with existing query, that is  `A -Relation1->B-Realtion2->A` , it only deletes if there is exact match for bidirectional relation, else it will not make any delete to single relations. And when you use single directional queries to read a relation it will show your deleted tables as well.
Similarly, in case of add, if we add with query `A -Relation1->B-Realtion2->A`, it will create duplicate relation if there was existing only one way relation. And this will ultimately cause a loop and you will never be able to delete any owner/bookmark which was added from Frontend UI before this MR .

Comment (dikshathakur3119) [reply metadata_service/proxy/neo4j_proxy.py:962]: @PaschalisDim let me know if it makes sense

Comment (PaschalisDim) [reply metadata_service/proxy/neo4j_proxy.py:962]: Hi @dikshathakur3119 
Apologies, I was not clear enough here.
So, before this MR let's assume that there is a user who owns and follows a table (table A).
So he will have one relation (usr{user_matcher})<-[rel:OWNER]-(resource{resource_matcher}) and one relation
(usr{user_matcher})-[rel:FOLLOW]->(resource{resource_matcher}).
Then you make this MR.
After we merge this MR and it goes to production,
when the user will try the tables that he owns, the amundsenmeta library will search for (resource{resource_matcher})-[r1:OWNER]->(usr{user_matcher})-[r2:OWNER_OF]->' \
                       f'(resource{resource_matcher}) 
these type of matches in order to see whether the user owns any table.
Since before this MR the user just had one-directional, the tableA will not show up in the list of the owned tables.
If he adds himself as an owner to another table, tableB, then he will have applied right the bi-directional and he will just see tableB in the list of owned tables.
To the point, for the existing one-way relations this change is a breaking change.
The add functionality should be fine. The get functionality is in my thoughts not right, since it will get just owners that have assigned themselves as owners just only AFTER this MR was deployed while I am not sure about the delete one which needs to be tested after the GET functionality is fixed.
@dikshathakur3119

**Comment by PaschalisDim** [comment]
@dikshathakur3119 
Another thing that came to my mind today is that we need to make sure that we also update the neo4j_staleness_removal_task from databuilder when this one gets finalized, to ensure deletion of all the possible relationships of the table, since after this MR there will be more relationships added.
Thanks!

File: tests/unit/proxy/test_neo4j_proxy.py
Comment (pPanda-beta) [inline tests/unit/proxy/test_neo4j_proxy.py:982]: @verdan 
We need something  similar (and precise as well) for atlas as well, what do you think?

**Pull request merged**
